### PR TITLE
dev-java/protobuf-java: add 4.28.0 | dev-java/jdbc-mysql: add 9.0.0

### PR DESCRIPTION
**Pull request by vaukai**
---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:19]:
```suggestion
SRC_URI="https://github.com/protocolbuffers/protobuf/releases/download/v${PV}/${P}.tar.gz"
```

The old one is a snapshot of the tag. That can change and might lead to trouble. I'm changing the rest of protobuf as well.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:29]:
```suggestion
	system-protoc? ( ~dev-libs/protobuf-${MY_PV4}:0[protoc] )
```
See https://bugs.gentoo.org/938102

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:8]:
```suggestion
JAVA_TESTING_FRAMEWORKS="junit-4"
TRUTH_PV="1.1.3"
```

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:20]:
```suggestion
	test? ( https://repo1.maven.org/maven2/com/google/truth/truth/${TRUTH_PV}/truth-${TRUTH_PV}.jar )"
```

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:124]:
```suggestion
	JAVA_GENTOO_CLASSPATH_EXTRA="${DISTDIR}/truth-${TRUTH_PV}.jar:testdata.jar"
```

File: profiles/base/package.use.mask
Comment (negril) [inline profiles/base/package.use.mask:14]: There will be versions in the tree before this gets merged.

**Review by negril — COMMENTED** [review]
You have a lot hard coded version strings in all ebuilds, that require you to update them every time -> easy to miss.

Link to the master branch or reference the file in `S`.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (vaukai) [reply dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:19]: saving the download as `4.28.0.tar.gz` instead of `28.0.tar.gz` would most probably cause a `MatchingChksums` complaint in ci (in case the corresponding `dev-libs/protobuf` got added)

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:18]:
```suggestion
PROTOBUF_PV="$(ver_cut 2-)"
```

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [reply dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:19]:
```suggestion
https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOBUF_PV}/${PN%%-java}-${PROTOBUF_PV}.tar.gz
```
I copied from the wrong file. Same reason though.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:21]:
```suggestion
S="${WORKDIR}/protobuf-${PROTOBUF_PV}"
```

**Comment by negril** [comment]
There is no reason to drop 4.27.2. It's a candidate for stabilization or we'll never get the backlog of old versions cleaned...

https://bugs.gentoo.org/938832

**Review by fordfrog — COMMENTED** [review]
just a question, as i'm kinda confused, why we don't rather stabilize related `profobuf`? i noticed there's a lot of work with supporting both the system-one and using the bundled one, i just don't know why it's this way.

**Comment by fordfrog** [comment]
@vaukai we had a chat with @negril over all the mess with `jdbc-mysql` needing the old `protobuf-java` etc... he found out that `jdbc-mysql-9.0` might not need the old version of `protobuf-java` anymore and we could instead use the latest one (see https://github.com/mysql/mysql-connector-j/commit/d3102b116434826e26bac7f0d177c71c1d7726c5#diff-f4bfdbd93e2caea722a038fca53c97874451c8eeb3ac6325b1981ce5d6eb6bd8L59-R67). if that would work, we could try to stabilize the new `jdbc-mysql-9.0` sooner and get rid of all the older versions of `protobuf-java`, which would simplify the maintenance i guess.

so, could you try to bump `jdbc-mysql` to version `9.0` and try it with the latest `protobuf-java`, or at least version `4.26.1` if the newest would not work?

**Comment by vaukai** [comment]
> just a question, as i'm kinda confused, why we don't rather stabilize related `profobuf`?

short version: we would have to wait until someone does `cc-arches` for related `dev-libs/protobuf`.
this way, we're independent.

**Comment by fordfrog** [comment]
> > so, could you try to bump `jdbc-mysql` to version `9.0`
> 
> i am having `jdbc-mysql-9.0.0` wip quite some while, still some trouble. and of course, i'd really like to toss all the older versions.

that's imo the way to go.

**Comment by vaukai** [comment]
> > > so, could you try to bump `jdbc-mysql` to version `9.0`
> > 
> > 
> > i am having `jdbc-mysql-9.0.0` wip quite some while, still some trouble. and of course, i'd really like to toss all the older versions.
> 
> that's imo the way to go.

first step: https://github.com/gentoo/gentoo/pull/38370

**Review by fordfrog — COMMENTED** [review]
we can't drop `3.25.3` as of now afaics:
```
dev-java/jdbc-mysql/jdbc-mysql-8.1.0-r1.ebuild: <dev-java/protobuf-java-3.26.0:0
dev-java/jdbc-mysql/jdbc-mysql-8.3.0.ebuild:    >=dev-java/protobuf-java-3.25:0 <dev-java/protobuf-java-3.26:0
profiles/base/package.use.mask:=dev-java/protobuf-java-3.23.0 system-protoc
```

**Review by fordfrog — COMMENTED** [review]
this one has unresolved reviews from @negril

**Comment by fordfrog** [comment]
and i merged the first commit, that dropping the old version

**Comment by vaukai** [comment]
> we can't drop `3.25.3` as of now afaics:
> 
> ```
> dev-java/jdbc-mysql/jdbc-mysql-8.1.0-r1.ebuild: <dev-java/protobuf-java-3.26.0:0
> dev-java/jdbc-mysql/jdbc-mysql-8.3.0.ebuild:    >=dev-java/protobuf-java-3.25:0 <dev-java/protobuf-java-3.26:0
> profiles/base/package.use.mask:=dev-java/protobuf-java-3.23.0 system-protoc
> ```

got it. sorry

**Comment by vaukai** [comment]
> and i merged the first commit, that dropping the old version

already rebased. will continue tomorrow to see what's still missing.

**Comment by fordfrog** [comment]
@vaukai what's the status here now?

**Comment by vaukai** [comment]
> @vaukai what's the status here now?

all ready to merge. first commit is needed for https://bugs.gentoo.org/938837

**Comment by fordfrog** [comment]
i'm asking because there are still some unresolved reviews from @negril 
image

**Comment by vaukai** [comment]
> i'm asking because there are still some unresolved reviews from @negril

found them, all closed.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:31]:
```suggestion
	system-protoc? ( dev-libs/protobuf:${MY_PV}[protoc] )
```
The `:0` SLOT serves no purpose here. The rc candidates have the same abi as the releases. So you can use the SLOT in the spec.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:32]: See https://github.com/gentoo/gentoo/blob/master/dev-libs/protobuf/protobuf-28.0.ebuild#L8-L12 and https://github.com/gentoo/gentoo/blob/master/dev-libs/protobuf/protobuf-28.0.ebuild#L53 on how to make this cleaner.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild]:
```suggestion
	if ! use system-protoc; then
		cmake_src_prepare
	fi
```

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild]:
```suggestion
	if ! use system-protoc; then
		cmake_src_configure
	fi
```

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (negril) [inline dev-java/protobuf-java/protobuf-java-4.28.0.ebuild]:
```suggestion
	if ! use system-protoc; then
		cmake_src_compile
	fi
```

**Review by negril — CHANGES_REQUESTED** [review]
Added some more.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (vaukai) [reply dev-java/protobuf-java/protobuf-java-4.28.0.ebuild:32]: That's too much of overkill here

**Comment by negril** [comment]
I am asking you again to **not** drop `4.27.2`. That and `3.25.3` are the two versions we can actually stabilize right now. 

You need:
- `dev-java/protobuf-java-3*` for `dev-java/jdbc-mysql-8*`
- `dev-java/protobuf-java-4*` for `dev-java/jdbc-mysql-9*`.

`protobuf-27.2` seems fairly stable. Where as `protobuf-28.0` has potentially breaking changes, so has `dev-cpp/abseil-cpp-20240722.0`. There is no guarantee when those will get stabilized.

And I'm not going to carry anything other than protobuf `4.28.0` `4.27.2` `3.25.4` `3.25.3` going forward. The rest needs to go.

File: dev-java/protobuf-java/protobuf-java-4.28.0.ebuild
Comment (vaukai) [reply dev-java/protobuf-java/protobuf-java-4.28.0.ebuild]: would have been glad to getting such proposal when starting with touching this package first.  will keep it for next  version in some few or more days.

**Comment by fordfrog** [comment]
@negril so, wrt `3.25.3`, are we gonna stabilize related `protobuf` or are we gonna mask `system-protoc` on related java package?

**Comment by negril** [comment]
`protobuf-25.3` should be stabilized. I'm going to disable the broken tests in the uncommon arches for abseil-cpp. Then get `abseil-cpp:0/20230802` stabilized.

**Comment by vaukai** [comment]
> `protobuf-25.3` should be stabilized. I'm going to disable the broken tests in the uncommon arches for abseil-cpp. Then get `abseil-cpp:0/20230802` stabilized.

once that's done, the [use.stable.mask](https://github.com/gentoo/gentoo/pull/37789/commits/40d68dbbc7bfc1cf89aa87a362e89cc928821d49) can be removed as usual. don't let it block https://bugs.gentoo.org/938837 any longer.

**Comment by vaukai** [comment]
stabilization bug for `dev-libs/protobuf-25.3` is https://bugs.gentoo.org/939501, waiting for `cc-arches`.

**Review by fordfrog — APPROVED** [review]

**Review by fordfrog — APPROVED** [review]

**Review by fordfrog — APPROVED** [review]

**Review by fordfrog — APPROVED** [review]

**Pull request closed without merging**
